STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_registry.c

```diff
@@ -0,0 +1,300 @@
+
+static az_result result_from_hal_status(HAL_StatusTypeDef status)
+{
+  switch (status)
+  {
+    case HAL_OK:
+      return AZ_OK;
+    case HAL_ERROR:
+      return AZ_ERROR_ULIB_SYSTEM;
+    case HAL_BUSY:
+      return AZ_ERROR_ULIB_BUSY;
+    case HAL_TIMEOUT:
+      return AZ_ERROR_ULIB_SYSTEM; // should be AZ_ERROR_ULIB_TIME_OUT
+    default:
+      return AZ_ERROR_ULIB_SYSTEM;
+  }
+}
+
+static az_result set_registry_node_ready_flag(uint8_t* address)
+{
+  AZ_ULIB_TRY
+  {
+    // if ((hal_status = internal_flash_write_doubleword((uint32_t)address, REGISTRY_READY)) !=
+    // HAL_OK)
+    // {
+    //   result = result_from_hal_status(hal_status);
+    //   AZ_ULIB_THROW_IF_AZ_ERROR(result);
+    // }
+    AZ_ULIB_THROW_IF_AZ_ERROR(
+        result_from_hal_status(internal_flash_write_doubleword((uint32_t)address, REGISTRY_READY)));
+  }
+  AZ_ULIB_CATCH(...) {}
+
+  return AZ_ULIB_TRY_RESULT;
+}
+
+static az_result set_registry_node_delete_flag(uint8_t* address)
+{
+  AZ_ULIB_TRY
+  {
+    // if ((hal_status = internal_flash_write_doubleword(
+    //          ((uint32_t)address + AZ_ULIB_REGISTRY_FLAG_SIZE), REGISTRY_DELETED))
+    //     != HAL_OK)
+    // {
+    //   result = result_from_hal_status(hal_status);
+    //   AZ_ULIB_THROW_IF_AZ_ERROR(result);
+    // }
+    AZ_ULIB_THROW_IF_AZ_ERROR(result_from_hal_status(internal_flash_write_doubleword(
+        ((uint32_t)address + AZ_ULIB_REGISTRY_FLAG_SIZE), REGISTRY_DELETED)));
+  }
+  AZ_ULIB_CATCH(...) {}
+
+  return AZ_ULIB_TRY_RESULT;
+}
+
+static az_result store_registry_node(az_ulib_registry_node node, uint8_t** node_ptr)
+{
+  AZ_ULIB_TRY
+  {
+    az_ulib_registry_node* runner;
+    int32_t node_index;
+
+    runner = (az_ulib_registry_node*)(&__REGISTRYINFO);
+
+    /* Look for next available spot */
+    for (node_index = 0; node_index < MAX_AZ_ULIB_REGISTRY_ENTRIES; node_index++)
+    {
+      if (runner->_internal.status.ready_flag == REGISTRY_FREE)
+      {
+        break;
+      }
+      runner++;
+    }
+
+    /* Handle case if all nodes were used */
+    AZ_ULIB_THROW_IF_ERROR((node_index < MAX_AZ_ULIB_REGISTRY_ENTRIES), AZ_ERROR_OUT_OF_MEMORY);
+
+    /* Store pointers to key value pair into flash */
+    AZ_ULIB_THROW_IF_AZ_ERROR(result_from_hal_status(internal_flash_write(
+        (uint8_t*)(runner) + sizeof(az_ulib_registry_node_status), // add size of flags
+        (uint8_t*)&(node._internal.data),
+        sizeof(az_ulib_registry_node_data))));
+
+    /* Return pointer to this node for setting flags later */
+    *node_ptr = (uint8_t*)runner;
+  }
+  AZ_ULIB_CATCH(...) {}
+
+  return AZ_ULIB_TRY_RESULT;
+}
+
+/* Find pointer to the next available space in flash to store registry key/value */
+static uint8_t* flash_buf_get()
+{
+  uint8_t* flash_ptr;
+  az_ulib_registry_node node;
+
+  /* Is there an empty registry? */
+  node = *(az_ulib_registry_node*)(&__REGISTRYINFO);
+  if (node._internal.status.ready_flag == REGISTRY_FREE)
```

Comment:
>if (node._internal.status.ready_flag == REGISTRY_FREE) [](http://example.com/codeflow?start=2&length=54)

You are assuming that if the first dword in the _REGISTRYINFO is 0xFFFFFFFFFFFFFFFF (REGISTRY_FREE), this is the first time that you write a registry. However, it is not necessarily true. For example, if you have a power off between lines 282 and 286, you will have an invalid registry in the first node, with a key without value and the first dword will still be set as 0xFFFFFFFFFFFFFFFF. Next time that you will write a register, the system will identify that the first node was compromised and will use the second node. In this case, even with the first node marked as REGISTRY_FREE, the second node contains a valid data. 

The correct approach here is run over all valid nodes to find the one with the biggest address.